server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceShardsAllocator.java

```diff
@@ -193,6 +200,28 @@ public void allocate(RoutingAllocation allocation, ActionListener<Void> listener
         reconcile(currentDesiredBalance, allocation);
     }
 
+    private void processNodeShutdowns(ClusterState clusterState) {
+        final var nodes = clusterState.nodes();
+        final var nodeShutdowns = clusterState.metadata().nodeShutdowns();
+        // If we remove a shutdown marker from a node, but it is still in the cluster, we'd need a reset.
+        boolean reset = processedNodeShutdowns.stream()
+            .anyMatch(nodeId -> nodeShutdowns.contains(nodeId) == false && nodes.get(nodeId) != null);
+        // Clean up processed shutdowns that are removed from the cluster metadata
+        processedNodeShutdowns.removeIf(nodeId -> nodeShutdowns.contains(nodeId) == false);
+
+        Set<String> newShutdowns = new HashSet<>();
+        for (var shutdown : nodeShutdowns.getAll().entrySet()) {
+            if (shutdown.getValue().getType() != SingleNodeShutdownMetadata.Type.RESTART
+                && processedNodeShutdowns.contains(shutdown.getKey()) == false) {
+                newShutdowns.add(shutdown.getKey());
+            }
+        }
```

Comment:
Might be simplified:

```suggestion
        for (var shutdown : nodeShutdowns.getAll().entrySet()) {
            if (shutdown.getValue().getType() != SingleNodeShutdownMetadata.Type.RESTART) {
                reset |= processedNodeShutdowns.add(shutdown.getKey());
            }
        }
```

this way we do not need to allocate additional collection

Reply:
I think in practice what we've ended up with effectively does a reset on almost all shutdown metadata changes, so I've made that explicit in 8bf34302565e1b918d208804709f71881c99fcfc. WDYT @idegtiarenko